### PR TITLE
Test: cts-fencing, cts-exec: Bail out if daemon is running

### PR DESCRIPTION
**Pull request by nrwahl2**
If `pacemaker-fenced` is already running when cts-fencing starts, `cts-fencing` will kill the existing fencer. Instead of allowing this to happen, we should bail out and tell the user to stop the cluster.

Likewise for `pacemaker-execd` (or `pacemaker-remoted`, if the `-R` option is used) and `cts-exec`.

Closes: [T496](https://projects.clusterlabs.org/T496)
Closes: [T497](https://projects.clusterlabs.org/T497)

### CODE REVIEW

**Comment by nrwahl2** [comment]
Should we check for `pacemakerd` or `corosync` running or is this sufficient? Probably fine as-is; all daemons ought to be up if any are.

File: cts/cts-fencing.in [hunk 1]
```diff
@@ -168,6 +169,24 @@ def killall(process):
     test.wait()
 
 
+def is_any_proc_running(process_names=[]):
```
Comment (nrwahl2) [inline cts/cts-fencing.in:172]: Currently not used directly. Planning to use the same scheme in `cts-exec`, where we do check for multiple processes.

It seems like pulling a lot stuff into a shared cts lib would be nice. In fact I think we've discussed that as a future low-priority goal.

**Comment by nrwahl2** [comment]
Network issue in CI.

**Comment by nrwahl2** [comment]
retest this please

File: cts/cts-exec.in [hunk 2]
```diff
@@ -1238,6 +1248,16 @@ def main(argv):
...
         sys.exit(CrmExit.OK)
 
+    if opts.options['pacemaker-remote']:
+        daemon_name = "pacemaker-remoted"
+    else:
+        daemon_name = "pacemaker-execd"
+
+    if is_proc_running(daemon_name):
+        print(f"Error: {daemon_name} is already running!")
+        print(f"Run {sys.argv[0]} only when the cluster is stopped.")
```
Comment (kgaillot) [inline cts/cts-exec.in:1258]: Pacemaker supports Python >=3.4, and sadly f strings were not introduced until 3.6

**Review by kgaillot — COMMENTED** [review]
Looks good. Did you decide against removing the killall code in case some test has unwanted leftovers?

File: cts/cts-exec.in (hunk 2 above)
Comment (nrwahl2) [reply cts/cts-exec.in:1258]: Darn, will fix.

**Comment by kgaillot** [comment]
> > Looks good. Did you decide against removing the killall code in case some test has unwanted leftovers?
> 
> Pretty much yeah. `killall(pacemakerd)` and `killall(pacemaker-fenced)` in `cts-fencing` (and their analogs in `cts-exec`) are run at the beginning of **each** test. That might still be legitimate, to make sure we don't have two copies running at a time, each one started by a test. They could all be replaced by `psutil` calls if we want. `psutil`'s `Process` object has `terminate()` and `kill()` methods.

It is a long-term goal to replace shell calls with Python native calls wherever possible, so it would be nice if straightforward but not worth a whole lot of time

> 
> `killall(corosync)` in `cleanup_environment()` seems legitimate, for killing the `corosync` that `cts-fencing` started.
> 
> `killall(corosync)` in `setup_environment` OTOH probably _should_ never get run. I guess it's nice and symmetrical relative to `cleanup_environment()`. Kind or related to the question "Should we check for `pacemakerd` or `corosync` running?"

I guess to be safe we should check for corosync running (if we will run the corosync tests).

**Comment by nrwahl2** [comment]
Two new commits:
* Added test for corosync running in cts-fencing if --no-cpg is not set.
* Use `psutil` to kill processes.

I didn't do this yet, but I'm inclined to say we should make a list of all pacemaker daemons plus corosync, and simply bail out if **any** of them are running (and give a list of which ones are running). We can find them all in one pass over `psutil.process_iter()`, making a list of the ones we found and including their names in the error message.

**Comment by kgaillot** [comment]
> I didn't do this yet, but I'm inclined to say we should make a list of all pacemaker daemons (plus corosync if we're using it), and simply bail out if **any** of them are running (and give a list of which ones are running). We can find them all in one pass over `psutil.process_iter()`, making a list of the ones we found and including their names in the error message.

Good idea, but not necessary for this PR. At some point I'd like to create a python3-pacemaker package (T255) for a module that can be imported by all the various tests, then we can reduce a lot of code duplication with some good functions like that.

File: cts/cts-exec.in [hunk 3]
```diff
@@ -102,6 +102,27 @@ def output_from_command(command):
...
+            procs.append(proc)
+
+    if terminate:
+        for proc in procs:
+            proc.terminate()
+        gone, alive = psutil.wait_procs(procs, timeout=3)
+        procs = alive
+
+    for proc in procs:
+        proc.kill()
```
Comment (kgaillot) [inline cts/cts-exec.in:123]: Minimizing TOCTOU risk is more important than efficiency when killing processes. You wouldn't think that there's enough time here for one of the processes to exit and a new one to reuse its pid, but it's just a matter of time before it surprises you. :) Better to check for and kill them one by one, even grabbing the entire process list again each time.

Of course there's also a chance that a relevant process will start after you check, but that's a less severe consequence and extremely unlikely if someone's running CTS tests.

**Review by kgaillot — COMMENTED** [review]
Minor nitpick

File: cts/cts-exec.in (hunk 3 above)
Comment (nrwahl2) [reply cts/cts-exec.in:123]: We should be good.

> Note: the way this class is bound to a process is uniquely via its PID. That means that if the process terminates and the OS reuses its PID you may end up interacting with another process. The only exceptions for which process identity is preemptively checked (via PID + creation time) is for the following methods: ... [kill()](https://psutil.readthedocs.io/en/latest/#psutil.Process.kill).

https://psutil.readthedocs.io/en/latest/#psutil.Process

> Kill the current process by using SIGKILL signal preemptively checking whether PID has been reused. 

https://psutil.readthedocs.io/en/latest/#psutil.Process.kill

The actual reuse check:
https://github.com/giampaolo/psutil/blob/379598312f60bf414afa8bf549f7f26af9e578ea/psutil/__init__.py#L571-L592
https://github.com/giampaolo/psutil/blob/379598312f60bf414afa8bf549f7f26af9e578ea/psutil/__init__.py#L266-L279

Comment (nrwahl2) [reply cts/cts-exec.in:123]: BTW (not directly relevant but as a note) I pretty much grabbed the "kill several processes" pattern from here:
- https://psutil.readthedocs.io/en/latest/#psutil.wait_procs

Comment (kgaillot) [reply cts/cts-exec.in:123]: Ah, that's great! They actually put some thought into that one :)

**Pull request merged**